src/libstd/thread/local.rs

```diff
@@ -293,13 +294,14 @@ mod imp {
             Key {
                 inner: UnsafeCell::new(None),
                 dtor_registered: Cell::new(false),
-                dtor_running: Cell::new(false)
             }
         }
 
         pub unsafe fn get(&'static self) -> Option<&'static UnsafeCell<Option<T>>> {
-            if intrinsics::needs_drop::<T>() && self.dtor_running.get() {
-                return None
+            if intrinsics::needs_drop::<T>() {
+                if DTOR_RUNNING.with(|running| { running.get() }) {
```

Comment:
The change I posted takes a pretty extreme stance: if we are destroying TLS, we are not guaranteed that any TLS is usable.
Conversely the failing test checks that never-used TLS variable can be initialised during the destruction of TLS.

Should we allow never-used TLS variables to be initialised while TLS is being destroyed?
This could be fixed by using the same approach in MacOS X as in Linux (i.e. subscribing a single destructor manager, then keeping a list of dtors to be run).

I'm not sure whether this guarantee is very convenient, as *only* the variables that have never been used are known to be useable. The other TLS variables available during the teardown of TLS depend on the order in which they are destroyed, which looks very fragile to me.